File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
n, m = map(int, input().split())

graph = [[0 for i in range(m)] for j in range(n)]
visited = [[0 for i in range(m)] for j in range(n)]
visited[0][0] = 1

for i in range(n):
    temp = input()
    for j in range(m):
        graph[i][j] = temp[j]

stack = []
stack.append([0,0,0])

dx = [0,0,-1,1]
dy = [1,-1,0,0]

answer = 0
while(stack):
    x, y, z = stack.pop(0)
    if x == n-1 and y == m-1:
        answer = z + 1
        break

    for i in range(4):
        if 0 <= dx[i] + x < n and 0 <= dy[i] + y < m and graph[dx[i] + x][dy[i] + y] == '1' and visited[dx[i] + x][dy[i] + y] == 0:
            stack.append([dx[i] + x , dy[i] + y, z+1])
            visited[dx[i] + x][dy[i] + y] = 1

print(answer)

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	static int[][] arr;
	static int n;
	static int m;
	static int[][] visited;
	static int[][] answer;
	static int[] dx = {0,0,-1,1};
	static int[] dy = {1,-1,0,0};
	
	public static void main(String[] args) throws IOException {
		
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(bf.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[n][m];
		visited = new int[n][m];
		answer = new int[n][m];
		
		for(int i=0; i<n; i++) {
			String s = bf.readLine();
			for(int j=0; j<m; j++)
				arr[i][j] = s.charAt(j) - 48;
		}
		
		Queue<int[]> queue = new LinkedList<>();
		queue.add(new int[] {0,0,1});
		visited[0][0] = 1;
		int answer = 0;
		
		while(!queue.isEmpty()) {
			int[] temp = queue.poll();
			int x = temp[0];
			int y = temp[1];
			
			if(x==n-1 && y==m-1)
				answer = temp[2];
			
			for(int i=0; i<4; i++) {
				int nx = dx[i] + x;
				int ny = dy[i] + y;
				
				if(nx>=0 && ny>=0 && nx<n && ny<m && visited[nx][ny]==0 && arr[nx][ny]==1) {
					queue.add(new int[] {nx, ny, temp[2]+1});
					visited[nx][ny] = 1;
				}
				
			}
		}
		
		System.out.println(answer);
	}

}